blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+  getBook,
+  getAuthorsOfBook,
+  getTagsOfBook,
+  (book, authors, tags): BookView =>
+    book && {
+      title: book.title,
+      description: book.description,
+      authors,
+      tags,
+    }
+)
+```
+
+And it will be used in the component and template like this:
+
+```ts
+book$ = this.store.pipe(select(getBookView, this.bookId));
+```
+
+```html
+<div *ngIf="book$ | async as book">
+  <div>{{ book.title }}</div>
+  <div><span *ngFor="let a of book.authors">{{ a.lastName }}</span></div>
+  <div>{{ book.description }}</div>
+  <div><span *ngFor="let t of book.tags">{{ t.name }}</span></div>
+</div>
+```
+
+This way of implementing it provides a good balance of advantages and disadvantages. We push the de-normalization further back so that it doesn't have to be handled in components and templates every time, That way code duplication is prevented. Also, the state still effectively mirrors the API and this simplifies managing and updating entities. The normalization in the data layer also makes it quite easy to later compose selectors for specific tasks, like:
+
+- search for books of a specific author
+- find books with a specific tag
+
+The composition of selectors can however affect memoization, because composed selectors fire every time their input changes. If those inputs are not correctly memoized, the resulting selector will also not be properly memoized.
+
+You can see this behavior in the [StackBlitz Example](https://stackblitz.com/github/dhhyi/ngrx-data-views/tree/basic-example?file=src%2Fapp%2Fstore%2Fbook-view%2Findex.ts). Whenever a Tag, Author or Book is updated, all Books will be subject to a view update.
+
+## The new problem with Memoization
+
+> In computing, memoization [...] is an optimization technique used primarily to speed up computer programs by storing the results of expensive function calls and returning the cached result when the same inputs occur again.
+[[Wikipedia]](https://en.wikipedia.org/wiki/Memoization)
+
+When using NgRx Store with Angular, selectors provide data with Observable streams. Whenever data is changing in the background, the view also has to be updated. As the state in NgRx is basically kept as one big structured object, every modification in reducers potentially triggers an update. That's why memoization is used to prevent pushing an update onto the Observable streams, when effectively nothing has changed. Keeping data in selectors properly memoized is the bread and butter of optimization. Whenever a selector fires unnecessary, it would initiate a needless and expensive view update.
+
+Let's have a look in the [official sources](https://github.com/ngrx/platform/blob/master/modules/store/src/selector.ts) how memoization works in the `createSelector` function. Every selector is created with `createSelectorFactory` and the argument for memoization, by default using `defaultMemoize`, which returns a `MemoizedProjection` for every selector doing the following:
```

Comment:
"Let's have a look at the official sources of NgRx to find out how memoization works...."